src/Compilers/Core/Portable/SourceGeneration/RunResults.cs

```diff
@@ -70,6 +70,17 @@ public ImmutableArray<SyntaxTree> GeneratedTrees
                 return _lazyGeneratedTrees;
             }
         }
+
+        public sealed class CancelledResult : GeneratorDriverRunResult
+        {
+            internal CancelledResult(TimeSpan elapsedTime, GeneratorRunResult? cancelledOn)
+                : base(ImmutableArray<GeneratorRunResult>.Empty, elapsedTime)
+            {
+                LastGeneratorRunning = cancelledOn;
+            }
+
+            public GeneratorRunResult? LastGeneratorRunning { get; }
```

Comment:
Given this is a GeneratorRunResult, is there data that's usable other than the time? Will there be whatever results _did_ come back up until the point of cancellation?

Reply:
No. The results are empty. This was one of the things I was really unsure about and why I ended up going with a subtype rather than a property on the parent class.

It gets pretty confusing if we return partial results as we either:
1. Only return the ones we actually ran
   - You get a variable number of results back, and the host has to track which ones came back compared to the last run
   - Host has to decide what to do with the 'missing' results.
2. We return cached results for the ones that didn't run:
   - You need to add more properties to the results, and have the host go through and check them. The cached values may not actually compile now, if the source compilation has changed significantly. 
   - That is true if we return nothing too, but at least *everything* is out of date, rather than some strange hybrid situation.